File: src/handler.rs
use warp::{Filter, Reply, Rejection};
use std::collections::BTreeMap;
use std::sync::Arc;
use google_cloud_spanner::client::{Client, ClientConfig, TxError};
use google_cloud_spanner::mutation::{insert_struct, update};
use google_cloud_spanner::statement::{Statement, ToStruct, Types, Kinds, ToKind};
use google_cloud_spanner::reader::{StatementReader, AsyncIterator, RowIterator};
use google_cloud_spanner::transaction::{ReadOptions, Transaction};
use google_cloud_spanner::row::{Error as RowError, TryFromStruct, Struct};
use google_cloud_googleapis::Status;
use google_cloud_spanner::transaction_ro::ReadOnlyTransaction;
use prost_types::{Value, Timestamp};
use google_cloud_spanner::value::CommitTimestamp;
use chrono::{Utc, TimeZone, NaiveDateTime, DateTime};
use uuid::Uuid;
use std::convert::Infallible;
use google_cloud_spanner::key::{Key, KeySet};
use crate::model;
use std::ops::DerefMut;
use google_cloud_gax::invoke::TryAs;
use crate::model::UserItem;
use google_cloud_spanner::sessions::SessionError;

#[derive(thiserror::Error, Debug)]
enum Error {
    #[error("no data found user_id = {0}")]
    NoDataFound(String),
    #[error(transparent)]
    ParseError(#[from] RowError),
    #[error(transparent)]
    GRPC(#[from] Status),
    #[error(transparent)]
    SessionError(#[from] SessionError),
}

impl TryAs<Status> for Error {
    fn try_as(&self) -> Result<&Status,()> {
        match self {
            Error::GRPC(s) => Ok(s),
            _ => Err(())
        }
    }
}

pub async fn create_user_handler(client: Client) -> Result<impl Reply, Rejection> {
    let mut ms = vec![];
    let user_id =  Uuid::new_v4().to_string();
    let user = model::User {
        user_id: user_id.to_string(),
        premium: true,
        updated_at: Utc::now(),
    };
    ms.push(insert_struct("User", user));
    for i in 0..10 {
        let user_item = model::UserItem {
            user_id: user_id.to_string(),
            item_id: i,
            quantity: 0,
            updated_at: Utc::now()
        };
        ms.push(insert_struct("UserItem", user_item));

        let user_character = model::UserCharacter{
            user_id: user_id.to_string(),
            character_id: i,
            level: 1,
            acquired_at: Utc::now(),
            updated_at: Utc::now()
        };
        ms.push(insert_struct("UserCharacter", user_character));
    }

    match client.apply(ms).await {
        Ok(_)  => {
            Ok(warp::reply::with_status(
                warp::reply::html(user_id.to_string()),
                warp::http::StatusCode::OK,
            ))
        },
        Err(e) => {
            log::error!("error {:?}", e);
            Ok(warp::reply::with_status(
                warp::reply::html(match e {
                    TxError::GRPC(e)  => e.to_string(),
                    TxError::InvalidSession(e) => "session error".to_string()
                }),
                warp::http::StatusCode::INTERNAL_SERVER_ERROR,
            ))
        }
    }
}

pub async fn update_inventory_handler(client: Client, user_id: String) -> Result<impl Reply, Rejection> {
    let tx_result = client.read_write_transaction(|mut tx| async {
        let result : Result<(), Error> = async {
            //TODO key 指定が必要
            let mut stmt = Statement::new("SELECT * From UserItem WHERE UserId = @UserId");
            stmt.add_param("UserId", user_id.to_string());
            let mut reader = tx.query(stmt).await?;
            let mut ms = vec![];
            loop {
                let row = reader.next().await?;
                match row {
                    Some(row) => {
                        let item_id = row.column_by_name::<i64>("ItemId")?;
                        let quantity = row.column_by_name::<i64>("Quantity")?;
                        ms.push(update("UserItem", vec!["UserId","ItemId","Quantity"], vec![
                            user_id.to_string().to_kind(),
                            item_id.to_kind(),
                            (quantity + 1).to_kind(),
                        ]));
                    },
                    None => break
                }
            };
            tx.buffer_write(ms);
            Ok(())
        }.await;
        (tx, result)
    }).await;

    match tx_result {
        Ok(success)  => {
            Ok(warp::reply::with_status(
                warp::reply::html(format!("ts={}", match success.0 {
                    Some(s) => s.seconds,
                    None => -1
                })),
                warp::http::StatusCode::OK,
            ))
        },
        Err(e) => {
            log::error!("error {:?}", e);
            Ok(warp::reply::with_status(
                warp::reply::html( format!("error {:?}",e)),
                warp::http::StatusCode::INTERNAL_SERVER_ERROR,
            ))
        }
    }
}

pub async fn read_inventory_handler(client: Client, user_id: String) -> Result<impl Reply, Rejection> {
    let mut tx = match client.single().await {
        Ok(tx) => tx,
        Err(e) => {
            let error_message =  "aaa".to_string();
            return Ok(warp::reply::with_status(
                warp::reply::html(error_message),
                warp::http::StatusCode::INTERNAL_SERVER_ERROR,
            ))
        }
    };
    match read(user_id, tx.deref_mut()).await {
        Ok(inventory)  => {
            Ok(warp::reply::with_status(
                warp::reply::html(format!("user={}, item={}, character={}", inventory.0, inventory.1,inventory.2)),
                warp::http::StatusCode::OK,
            ))
        },
        Err(e) => {
            log::error!("error {:?}", e);
            Ok(warp::reply::with_status(
                warp::reply::html( format!("error {:?}",e)),
                warp::http::StatusCode::INTERNAL_SERVER_ERROR,
            ))
        }
    }
}

async fn read(user_id:String, mut tx: &mut Transaction) -> Result<(String,usize,usize), Error> {
    let mut stmt = Statement::new("SELECT * , \
            ARRAY (SELECT AS STRUCT * FROM UserItem WHERE UserId = @Param1 ) AS UserItem, \
            ARRAY (SELECT AS STRUCT * FROM UserCharacter WHERE UserId = @Param1 ) AS UserCharacter  \
            FROM User \
            WHERE UserId = @Param1");
    stmt.add_param("Param1", user_id.to_string());
    let mut reader = tx.query(stmt).await?;
    let row = match reader.next().await?{
        Some(row) => row,
        None => return Err(Error::NoDataFound(user_id))
    };
    let user_id= row.column_by_name::<String>("UserId")?;
    let user_items= row.column_by_name::<Vec<model::UserItem>>("UserItem")?;
    let user_characters = row.column_by_name::<Vec<model::UserCharacter>>("UserCharacter")?;
    Ok((
        user_id,
        user_items.len(),
        user_characters.len()
    ))
}

File: spanner/go/cmd/internal/handler.go
package internal

import (
	"cloud.google.com/go/spanner"
	"context"
	"fmt"
	"github.com/google/uuid"
	"net/http"
	"time"
)

func errorResponse(writer http.ResponseWriter, err error ) {
	fmt.Printf("err %+v\n", err)
	writer.WriteHeader(http.StatusInternalServerError)
	_, _ = writer.Write([]byte(err.Error()))
}

func CreateUser(client *spanner.Client) http.HandlerFunc {
	return func(writer http.ResponseWriter, request *http.Request) {
		userID, err := uuid.NewRandom()
		if err != nil {
			errorResponse(writer,err)
			return
		}
		userIDString := userID.String()
		userMutation, err := spanner.InsertStruct("User", &User{
			UserId:    userIDString,
			Premium:   false,
			UpdatedAt: spanner.CommitTimestamp,
		})
		if err != nil {
			errorResponse(writer,err)
			return
		}
		ms := make([]*spanner.Mutation, 0)
		ms = append(ms, userMutation)

		for i := 0; i < 10; i ++{
			itemMutation, err := spanner.InsertStruct("UserItem", &UserItem {
				UserId:    userIDString,
				ItemId: int64(i),
				Quantity: 0,
				UpdatedAt: spanner.CommitTimestamp,
			})
			if err != nil {
				errorResponse(writer,err)
				return
			}
			ms = append(ms,itemMutation)

			characterMutation, err := spanner.InsertStruct("UserCharacter", &UserCharacter {
				UserId:    userIDString,
				CharacterId: int64(i),
				Level: 1,
				AcquiredAt: time.Now(),
				UpdatedAt: spanner.CommitTimestamp,
			})
			if err != nil {
				errorResponse(writer,err)
				return
			}
			ms = append(ms,characterMutation)
		}

		_, err = client.Apply(request.Context(),ms )
		if err != nil {
			errorResponse(writer, err)
		}
		writer.WriteHeader(http.StatusOK)
		_, _ = writer.Write([]byte(userIDString))
	}
}

func ReadInventory(client *spanner.Client) http.HandlerFunc {
	return func(writer http.ResponseWriter, request *http.Request) {
		userID := request.FormValue("user_id")
		stmt := spanner.NewStatement(`SELECT * ,
			ARRAY (SELECT AS STRUCT * FROM UserItem WHERE UserId = @Param1 ) AS UserItem, 
			ARRAY (SELECT AS STRUCT * FROM UserCharacter WHERE UserId = @Param1 ) AS UserCharacter  
			FROM User 
			WHERE UserId = @Param1
		`)
		stmt.Params["Param1"] = userID
		tx := client.Single()
		iter := tx.Query(request.Context(), stmt)
		defer iter.Stop()
		row, err := iter.Next()
		if err != nil {
			errorResponse(writer,err)
			return
		}
		var userId string
		if err = row.ColumnByName("UserId", &userId); err != nil {
			errorResponse(writer,err)
			return
		}
		var userItems []*UserItem
		if err = row.ColumnByName("UserItem", &userItems); err != nil {
			errorResponse(writer,err)
			return
		}

		var userCharacters []*UserCharacter
		if err = row.ColumnByName("UserCharacter", &userCharacters); err != nil {
			errorResponse(writer,err)
			return
		}
		writer.WriteHeader(http.StatusOK)
		_, _ = writer.Write([]byte(fmt.Sprintf("user=%d, item=%d, character=%d", userId, len(userItems), len(userCharacters))))
	}
}

func UpdateInventory(client *spanner.Client) http.HandlerFunc {
	return func(writer http.ResponseWriter, request *http.Request) {
		userID := request.FormValue("user_id")

		cts, err := client.ReadWriteTransaction(request.Context(), func(ctx context.Context, transaction *spanner.ReadWriteTransaction) error {
			stmt := spanner.NewStatement("SELECT * From UserItem WHERE UserId = @UserId")
			stmt.Params["UserId"] = userID
			iter := transaction.Query(ctx, stmt)
			ms := make([]*spanner.Mutation,0)
			for {
				row ,err := iter.Next()
				if err != nil {
					break
				}
				var itemId int64
				if err = row.ColumnByName("ItemId", &itemId); err != nil {
					return err
				}
				var quantity int64
				if err = row.ColumnByName("Quantity", &quantity); err != nil {
					return err
				}
				ms = append(ms, spanner.Update("UserItem", []string{"UserId", "ItemId", "Quantity"}, []interface{}{userID, itemId, quantity + 1}))
			}
			if err := transaction.BufferWrite(ms); err != nil {
				return err
			}
			return nil
		})

		if err != nil {
			errorResponse(writer,err)
			return
		}
		writer.WriteHeader(http.StatusOK)
		_, _ = writer.Write([]byte(fmt.Sprintf("ts=%d", cts.Second())))
	}
}
